fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSession.java

```diff
@@ -89,85 +115,267 @@ public void write(final String subpath, final InputStream stream) {
+                    .getVersionId()
+                    .toString();
+        } else {
+            // perform commit to head version
+            return ocflRepository.stageChanges(ObjectVersionId.head(objectIdentifier), commitInfo, updater -> {
+                updater.addPath(stagingPath, "", MOVE_SOURCE);
+            }).getVersionId().toString();
+        }
+    }
+
+    private String commitUpdates(final CommitOption commitOption) {
+        // Updater which pushes all updated files and then performs queued deletes
+        final Consumer<OcflObjectUpdater> commitChangeUpdater = updater -> {
+            if (!isStagingEmpty()) {
+                updater.addPath(stagingPath, "", MOVE_SOURCE, OVERWRITE);
+            }
+            deletePaths.forEach(updater::removeFile);
+        };
+
+        if (NEW_VERSION.equals(commitOption)) {
+            // check if a mutable head exists for this object
+            if (ocflRepository.hasStagedChanges(objectIdentifier)) {
+                // Persist the current changes to the mutable head, and then commit the head as a version
+                ocflRepository.stageChanges(ObjectVersionId.head(objectIdentifier),
+                        new CommitInfo(),
+                        commitChangeUpdater);
+                return ocflRepository.commitStagedChanges(objectIdentifier, new CommitInfo())
+                        .getVersionId().toString();
             } else {
-                throw new IllegalArgumentException("Invalid commit option provided: " + commitOption);
+                // Commit directly to a new version
+                return ocflRepository.updateObject(ObjectVersionId.head(objectIdentifier),
+                        new CommitInfo(),
+                        commitChangeUpdater)
+                        .getVersionId().toString();
             }
+        } else {
+            // perform commit to mutable head version
+            return ocflRepository.stageChanges(ObjectVersionId.head(objectIdentifier),
+                    new CommitInfo(),
+                    commitChangeUpdater)
+                    .getVersionId().toString();
         }
     }
 
+    /**
+     * {@inheritDoc}
+     *
+     * Close the session and perform cleanup.
+     */
+    @Override
+    public synchronized void close() throws PersistentStorageException {
+        sessionClosed = true;
+
+        cleanupStaging();
+    }
+
+    private boolean isStagingEmpty() {
+        return stagingPath.toFile().listFiles().length == 0;
+    }
+
+    private boolean hasStagedChanges(final Path path) {
+        return path.toFile().exists();
+    }
+
     private boolean isNewObject() {
-        return false;
+        return !ocflRepository.containsObject(objectIdentifier);
     }
 
     private Path resolveStagedPath(final String subpath) {
         return stagingPath.resolve(subpath);
     }
+
+    private synchronized void assertSessionOpen() throws PersistentSessionClosedException {
```

Comment:
I'm not sure this needs to be `synchronized`. Why is it a problem if it's called concurrently?